corehq/apps/couch_sql_migration/couchsqlmigration.py

```diff
@@ -391,6 +394,100 @@ def _process_missing_forms(self):
+                            yield form_id
+                elif jdiff.diff_type == 'set_mismatch' and jdiff.path[0] == 'xform_ids':
+                    yield from old_value.split(",")
+            elif (
+                kind == "stock state"
+                and isinstance(old_value, dict)
+                and "ledger" in old_value
+                and old_value.get("form_state") != MISSING_BLOB_PRESENT
+            ):
+                yield old_value["ledger"]["last_modified_form_id"]
+
+        from .missingdocs import MissingIds
+        iter_blob_forms = _get_missing_blob_present_form_loader(self.domain)
+        drop_sql_ids = MissingIds.forms(self.statedb).drop_sql_ids
+        for diff in _iter_case_diffs(self.statedb, self.stopper):
+            case_id = diff.doc_id
+            for form in iter_blob_forms(diff):
+                log.info("migrating form %s received on %s from case %s",
+                    form.form_id, form.received_on, case_id)
+                self._migrate_form(form, get_case_ids(form))
+            form_ids = set(iter_form_ids(diff.json_diff, diff.kind))
+            if not form_ids:
+                continue
+            missing_ids = set(drop_sql_ids(form_ids))
+            for form_id in missing_ids:
+                form = XFormInstance.get(form_id)
+                self._migrate_form(form, get_case_ids(form))
+                log.info("migrated missing form %s from case %s", form_id, case_id)
+            self._migrate_missing_cases_and_ledgers(form_ids - missing_ids)
+
+    def _migrate_missing_cases_and_ledgers(self, form_ids):
+        """Update cases and ledgers for forms that have already been migrated
+
+        :returns: a set of form ids that had cases or ledgers to migrate.
+        """
+        migrated_ids = set()
+        for form_id in form_ids:
+            saved = self._save_missing_cases_and_ledgers(form_id)
+            if saved:
+                migrated_ids.add(form_id)
+        return migrated_ids
+
+    def _save_missing_cases_and_ledgers(self, form_id):
+        def iter_missing_ledgers(stock_result):
+            assert not stock_result.models_to_delete, (form_id, stock_result)
+            if not (stock_result and stock_result.models_to_save):
+                return []
+            get_transactions = LedgerAccessorSQL.get_ledger_transactions_for_form
+            case_ids = {v.case_id for v in stock_result.models_to_save}
+            refs = {t.ledger_reference for t in get_transactions(form_id, case_ids)}
+            for value in stock_result.models_to_save:
+                if value.ledger_reference not in refs:
+                    yield value
+
+        from django.db import transaction
+        couch_form = XFormInstance.get(form_id)
+        sql_form = FormAccessorSQL.get_form(form_id)
+        result = self._get_case_stock_result(sql_form, couch_form)
+        if not result:
+            return False
+        cases = [c for c in result.case_models if c.has_tracked_models()]
```

Comment:
This is a little scary. I'm not sure if it is always safe to save all updates ("tracked models") returned by [`self._get_case_stock_result()`](https://github.com/dimagi/commcare-hq/blob/129e08a152022dd49f6404c76a9682c4bb8a6cfa/corehq/apps/couch_sql_migration/couchsqlmigration.py#L253-L266). Will it matter if an update is applied to a case twice? Will that create a new case transaction, or will it update the existing transaction? Is there a way to tell if changes present on a given "tracked model" have already been applied (and skip them if yes)?

Reply:
@snopoke I'd be interested in your feedback here.